url61db808cfe_ctriposs_bigcache_tgz-pJ8-com_ctriposs_bigcache_BigCacheStressTestJ8/src/com/ctriposs/bigcache/BigCache.java

```diff
@@ -0,0 +1,456 @@
+	}
+	
+	public long count(){
+		return pointerMap.size();
+	}
+
+    /**
+     * Get the latest stats of the cache.
+     *
+     * @return all stats.
+     */
+    public BigCacheStats getStats() {
+        return new BigCacheStats(hitCounter.get(), missCounter.get(), getCounter.get(),
+                putCounter.get(), deleteCounter.get(), purgeCounter.get(), moveCounter.get(),
+                count(), storageManager.getUsed(), storageManager.getDirty(),
+                storageManager.getCapacity(), storageManager.getUsedBlockCount(), storageManager.getFreeBlockCount(),
+                storageManager.getTotalBlockCount());
+    }
+
+    abstract static class CacheDaemonWorker<K> implements Runnable {
+        private WeakReference<BigCache<K>> cacheHolder;
+        private ScheduledExecutorService ses;
+
+        CacheDaemonWorker(BigCache<K> cache) {
+            ses = cache.ses;
+            cacheHolder = new WeakReference<BigCache<K>>(cache);
+        }
+
+        @Override
+        public void run() {
+            BigCache cache = cacheHolder.get();
+            if (cache == null) {
+                // cache is recycled abnormally
+                if (ses != null) {
+                    ses.shutdownNow();
+                    ses = null;
+                }
+                return;
+            }
+            try {
+                process(cache);
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+
+            cache.storageManager.clean();
+        }
+
+        abstract void process(BigCache<K> cache) throws IOException;
+    }
+
+    /**
+     * Clean the expired keys.
+     *
+     * @param <K>
+     */
+    static class CacheCleaner<K> extends CacheDaemonWorker<K> {
+        CacheCleaner(BigCache<K> cache) {
+            super(cache);
+        }
+
+        @Override
+        public void process(BigCache<K> cache) throws IOException {
+            Set<K> keys = cache.pointerMap.keySet();
+
+            // store the expired keys according to their associated lock
+            Map<ReadWriteLock, List<K>> expiredKeys = new HashMap<ReadWriteLock, List<K>>();
+
+            // find all the keys that may be expired. It's lock less as we will validate later.
+            for(K key : keys) {
+                CacheValueWrapper wrapper = cache.pointerMap.get(key);
+                if (wrapper != null && wrapper.isExpired()) {
+                    ReadWriteLock lock = cache.getLock(key);
+                    List<K> keyList = expiredKeys.get(lock);
+                    if (keyList == null) {
+                        keyList = new ArrayList<K>();
+                        expiredKeys.put(lock, keyList);
+                    }
+                    keyList.add(key);
+                }
+            }
+
+            // expire keys with write lock, this will complete quickly.
+            for (ReadWriteLock lock : expiredKeys.keySet()) {
+                List<K> keyList = expiredKeys.get(lock);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.